### PR TITLE
ARROW-17142: [Python] Parquet FileMetadata.equals() method segfaults when passed None

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes : https://issues.apache.org/jira/browse/ARROW-17142

### CODE REVIEW

File: python/pyarrow/_parquet.pyx [hunk 1]
```diff
@@ -680,6 +680,9 @@ cdef class FileMetaData(_Weakrefable):
         -------
         are_equal : bool
         """
+        err_msg = lambda: ("Expected `other` to be of type FileMetaData"
+                           " but found " + str(type(other)))
```
Comment (kshitij12345) [inline python/pyarrow/_parquet.pyx:684]: Wrap it in lambda so that we construct this only when the assertion fails.

**Comment by kshitij12345** [comment]
cc: @AlenkaF @jorisvandenbossche

File: python/pyarrow/_parquet.pyx (hunk 1 above)
Comment (lidavidm) [reply python/pyarrow/_parquet.pyx:684]: Why not just `if not isinstance…` and a regular `raise`? (assertions can technically be disabled.) And the idiomatic exception would be TypeError.

Finally, I think Cython already enforces the type due to the annotation above, so it's just that we need to check for None.

Comment (jorisvandenbossche) [reply python/pyarrow/_parquet.pyx:684]: > Finally, I think Cython already enforces the type due to the annotation above, so it's just that we need to check for None.

Indeed, and we can actually also let cython check for the None by changing the signature to `def equals(self, FileMetaData other not None):`

Comment (kshitij12345) [reply python/pyarrow/_parquet.pyx]: > def equals(self, FileMetaData other not None):

I didn't know that. Thanks! Will try that and update!

Comment (kshitij12345) [reply python/pyarrow/_parquet.pyx]: Updated :)

**Comment by kshitij12345** [comment]
I think this should be ready. Mac CI failure looks irrelevant :)

**Review by AlenkaF — APPROVED** [review]
LGTM! Thank you for reporting and fixing the issue!

**Review by lidavidm — APPROVED** [review]

**Pull request merged**